File: cpp/classic/dynamic_programming/KnapsackOfComplete.cpp
#include <iostream>
#include <vector>
using namespace std;

class KnapsackOfComplete {
public:
    int solution(vector<int> &weights, vector<int> &values, int target) {
        vector<int> result(target + 1, 0);
        for (int i = 0; i < (int) weights.size(); ++i) {
            int weight = weights[i];
            int value = values[i];
            for (int j = weight; j <= target; ++j) {
                int pivot = result[j - weight] + value;
                if (result[j] < pivot) {
                    result[j] = pivot;
                }
            }
        }
        return result.back();
    }
};

File: py/classic/dynamic_programming/KnapsackOfComplete.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class KnapsackOfComplete:
    def solution(self, weights: List[int], values: List[int], target: int) -> int:
        result = [0 for _ in range(target + 1)]
        for weight, value in zip(weights, values):
            for i in range(weight, target + 1):
                pivot = result[i - weight] + value
                if result[i] < pivot:
                    result[i] = pivot
        return result[-1]
